# Basic Algorithm Experiment

The purpose of this notebook is to evaluate the performance of basic algorithms. We will evaluate: 
- Breadth First Search
- Depth First Search
- Dijkstra's algorithm (adjusted to stop once a path has been found)


In [1]:
# correct working directory.
# This is necessary for imports because the notebook is not in the main folder of the project. 
if not "working_directory_corrected" in vars():
    %cd ..
    working_directory_corrected = True


import pandas as pd

from evaluation.timed_experiment import Timed_Experiment
from algorithms.depth_first_search import Depth_First_Search
from algorithms.breadth_first_search import Breadth_First_Search
from algorithms.dijkstra import Dijkstra

# load dataset
from data.dataset import Dataset
dataset = Dataset()
graph = dataset.load_graph()

c:\Users\frank\Documents\Teaching\LU\Planning and Optimization LU - Material\Planning Example Project\planning_example_project


c:\Users\frank\Documents\Teaching\LU\Planning and Optimization LU - Material\Planning Example Project\planning_example_project\environment\lib\site-packages\IPython\core\magics\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


## Procedure

For this experiment we will run Breadth First Search, Depth First Search and Dijkstra's Algorithm.

Each algorithm will be tested on 1000 planning problems (meaning different start and end nodes). To eliminate differences due to randomness, the three algorithms will be synchronized based on a random seed, meaning they will solve the same planning problems. 

From each experiment we will collect the following data:
- *average_time*: The average time needed to find a path.
- *nr_extended*: The number of nodes that were extended during search.
- *time per extension*: The average time required for one extension.

The experiment will be run using the class *Timed_Experiment*.



## Results

The below cell runs the experiment and prints the results.

If you want to run your own experiments you can change this cell. If you want to convince yourself that the three algorithms are indeed tested on the same problems, you can set the verbose flag to True.

In [5]:
algorithms = [Breadth_First_Search, Depth_First_Search, Dijkstra]

for algorithm in algorithms:
    print(f"Running {algorithm.__name__}...")
    experiment = Timed_Experiment(graph, algorithm, 1000, random_seed=42, verbose = False)
    experiment.run()
    print("Average Time: ", int(experiment.get_average_time()), "ns")
    print("Nr Extended: ", experiment.get_nr_extensions())
    print("Time per extension: ", int(experiment.get_average_extension_time()), "ns")
    print()



Running Breadth_First_Search...
Average Time:  101199413 ns
Nr Extended:  33751457
Time per extension:  2998 ns

Running Depth_First_Search...
Average Time:  76471816 ns
Nr Extended:  33502919
Time per extension:  2282 ns

Running Dijkstra...
Average Time:  205628019 ns
Nr Extended:  33867838
Time per extension:  6071 ns



The below table is the result of running the experiment on a Lenovo ThinkPad E14 Gen 6, standard configuration.
While running, the notebook has been left alone, to minimize the influence of other applications.


|          | Average Time | Nr Extended | Time per Extension |
| -------- | ------- | ------- | ------- |
| BFS| 101 ms | 33.751.457 | 2998     ns |
| DFS | 76 ms | 33.502.919 | 2282     ns |
| Dijkstra | 206 ms | 33.867.838 | 6071   ns|


# Discussion

The results indicate that Depth First Search is the fastest of the three alternatives by a considerable margin. 

Breadth First Search comes in second with about 20% longer average time.

Dijkstra's algorithm takes almost three times as long as Depth First Search.

The difference in times seem to be influenced most by the time per execution. Dikstra's algorithm takes almost three times as long as the other two algorithms. This is likely due to the need to maintain a priority queue, which has non-constant insertion complexity. 
